File: src/workers/revenue_scraper.py
import json
import os
import re
import urllib.request
import urllib.parse
from datetime import datetime

# Configurações Iniciais
OUTPUT_DIR = "squads/nexus/data"
RAW_FILE = os.path.join(OUTPUT_DIR, "raw_opportunities.json")
WORKANA_URL = "https://www.workana.com/jobs?category=it-programming"
NINE_NINE_URL = "https://www.99freelas.com.br/projects?categoria=web-desenvolvimento"

def setup():
    os.makedirs(OUTPUT_DIR, exist_ok=True)

def fetch_url(url, headers=None):
    if not headers:
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'}
    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req) as response:
            return response.read().decode('utf-8')
    except Exception as e:
        print(f"Erro ao buscar {url}: {e}")
        return None

def parse_reward(reward_str):
    try:
        numbers = [int(re.sub(r'\D', '', n)) for n in re.findall(r'\d+', reward_str.replace('.', ''))]
        if len(numbers) == 1:
            return 0, numbers[0]
        elif len(numbers) >= 2:
            return numbers[0], numbers[1]
    except:
        pass
    return 0, 0

def fetch_workana():
    projects = []
    print("\nBuscando Workana (se houver perfil logado ou não bloqueado)...")
    html = fetch_url(WORKANA_URL)
    if not html: return projects
    
    project_blocks = re.findall(r'<div class="project-item[^>]*>(.*?)</div>\s*</div>\s*</div>', html, re.DOTALL)
    for block in project_blocks[:5]: # Limit for speed
        try:
            title_match = re.search(r'<h2 class="project-title[^>]*>.*?<a href="([^"]+)"[^>]*title="([^"]+)"', block, re.DOTALL)
            if not title_match: continue
            
            relative_link = title_match.group(1)
            full_link = "https://www.workana.com" + relative_link
            title = title_match.group(2).strip()
            
            reward_match = re.search(r'<span class="values[^>]*>(.*?)</span>', block)
            reward_raw = reward_match.group(1).strip() if reward_match else "N/A"
            reward_min, reward_max = parse_reward(reward_raw)
            
            detailed_html = fetch_url(full_link)
            full_description = ""
            skills = []
            
            if detailed_html:
                desc_match = re.search(r'<div class="expander js-expander-body[^>]*>(.*?)</div>', detailed_html, re.DOTALL)
                if desc_match:
                    full_description = re.sub(r'<[^>]+>', '', desc_match.group(1)).strip()
                skills_matches = re.findall(r'<a class="skill[^>]*>(.*?)</a>', detailed_html)
                skills = [re.sub(r'<[^>]+>', '', s).strip() for s in skills_matches]

            projects.append({
                "id": f"wkn_{hash(title)}",
                "platform": "Workana",
                "title": title,
                "url": full_link,
                "rewardRaw": reward_raw,
                "rewardMin": reward_min,
                "rewardMax": reward_max,
                "full_description": full_description,
                "skills": skills,
                "status": "raw",
                "scraped_at": datetime.now().isoformat()
            })
            print(f"  [Workana] {title[:30]}...")
        except Exception as e:
            continue
    return projects

def fetch_99freelas():
    projects = []
    print("\nBuscando 99Freelas...")
    html = fetch_url(NINE_NINE_URL)
    if not html: return projects
    
    # Nova regex multi-linha para 99freelas
    matches = re.findall(r'<h1 class="title">\s*<a href="(/project/[^"]+)"[^>]*>([^<]+)</a>', html)
    
    for match in matches[:8]: # Pega as ultimas 8
        try:
            relative_link = match[0]
            # Replace HTML entities no title
            title = match[1].replace('&ccedil;', 'ç').replace('&atilde;', 'ã').replace('&iacute;', 'í').replace('&ecirc;', 'ê').replace('&otilde;', 'õ').strip()
            
            full_link = "https://www.99freelas.com.br" + relative_link
            
            detailed_html = fetch_url(full_link)
            full_description = ""
            skills = []
            reward_raw = "N/A"
            reward_min, reward_max = 0, 0
            
            if detailed_html:
                desc_match = re.search(r'<div class="text-keep-lines"(.*?)>(.*?)</div>', detailed_html, re.DOTALL)
                if desc_match:
                    full_description = re.sub(r'<[^>]+>', '', desc_match.group(2)).strip()
                
                skills_match = re.search(r'<p class="skills">Habilidades desejadas:(.*?)</p>', detailed_html, re.DOTALL)
                if skills_match:
                    skills_html = skills_match.group(1)
                    skills_list = re.findall(r'<a[^>]*>(.*?)</a>', skills_html)
                    skills = [s.strip() for s in skills_list]
                    
                reward_m = re.search(r'<b>Orçamento[a-zA-Z\s]*:</b>\s*(R\$[\s\d,\.]+ - R\$[\s\d,\.]+)', detailed_html)
                if reward_m:
                    reward_raw = reward_m.group(1).strip()
                    reward_min, reward_max = parse_reward(reward_raw)
            
            projects.append({
                "id": f"99f_{hash(title)}",
                "platform": "99Freelas",
                "title": title,
                "url": full_link,
                "rewardRaw": reward_raw,
                "rewardMin": reward_min,
                "rewardMax": reward_max,
                "full_description": full_description,
                "skills": skills,
                "status": "raw",
                "scraped_at": datetime.now().isoformat()
            })
            print(f"  [99Freelas] {title[:30]}...")
        except Exception as e:
            continue
            
    return projects

def run():
    print(f"Iniciando Varredura Multi-Plataforma (Revenue Scraper v3.1)...")
    setup()
    
    all_projects = []
    
    nine_projs = fetch_99freelas()
    all_projects.extend(nine_projs)
    
    workana_projs = fetch_workana()
    all_projects.extend(workana_projs)
    
    print(f"\nTotal: {len(all_projects)} oportunidades consolidadas.")
    
    if all_projects:
        with open(RAW_FILE, 'w') as f:
            json.dump(all_projects, f, indent=4, ensure_ascii=False)
        print(f"Salvo em: {RAW_FILE}")
    else:
        print("Falha ao coletar qualquer oportunidade.")

if __name__ == "__main__":
    run()

File: src/workers/revenue_scraper.js
const puppeteer = require('puppeteer');
const cheerio = require('cheerio');
const fs = require('fs');
const path = require('path');

const OUTPUT_DIR = path.join(__dirname, '../../squads/nexus/data');
const RAW_FILE = path.join(OUTPUT_DIR, 'raw_opportunities.json');
const WORKANA_URL = 'https://www.workana.com/jobs?category=it-programming';

async function runScraper() {
    console.log('Iniciando Varredura CLI-First (Puppeteer + Cheerio Deep Fetch)...');

    if (!fs.existsSync(OUTPUT_DIR)) {
        fs.mkdirSync(OUTPUT_DIR, { recursive: true });
    }

    const browser = await puppeteer.launch({
        headless: 'new',
        args: ['--no-sandbox', '--disable-setuid-sandbox']
    });

    // Lista de Links
    const page = await browser.newPage();
    await page.setUserAgent('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
    console.log(`Buscando projetos em: ${WORKANA_URL}`);
    await page.goto(WORKANA_URL, { waitUntil: 'networkidle2' });

    const htmlList = await page.content();
    const $list = cheerio.load(htmlList);
    const projectLinks = [];

    $list('.project-item').each((i, el) => {
        const item = $list(el);
        const relativeLink = item.find('.project-title a').attr('href');

        if (relativeLink) {
            const url = "https://www.workana.com" + relativeLink;
            const title = item.find('.project-title a').attr('title') || item.find('.project-title').text().trim();
            const rewardRaw = item.find('.values').text().trim() || 'N/A';

            const skills = [];
            item.find('.skills .skill').each((idx, s) => {
                skills.push($list(s).text().trim());
            });

            projectLinks.push({ url, title, rewardRaw, skills });
        }
    });

    console.log(`Encontrados ${projectLinks.length} links. Iniciando Extração Profunda (Deep Fetch)...`);
    const projects = [];

    for (let i = 0; i < Math.min(projectLinks.length, 10); i++) {
        const projData = projectLinks[i];
        console.log(`⬇️ Extraindo [${i + 1}/10]: ${projData.url}`);

        try {
            const detailPage = await browser.newPage();
            await detailPage.setUserAgent('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36');
            await detailPage.goto(projData.url, { waitUntil: 'domcontentloaded' });

            const detailHtml = await detailPage.content();
            const $ = cheerio.load(detailHtml);

            // Descrição Profunda
            const full_description = $('.expander').first().text().trim() || '';

            // Parse Min/Max Reward da string original
            const numbers = (projData.rewardRaw.match(/\d+(?:[.,]\d+)?/g) || []).map(n => parseInt(n.replace(/\D/g, '')));
            let rewardMin = 0, rewardMax = 0;
            if (numbers.length === 1) {
                rewardMax = numbers[0];
            } else if (numbers.length >= 2) {
                rewardMin = numbers[0];
                rewardMax = numbers[1];
            }

            // ID
            let hash = 0;
            for (let j = 0; j < projData.url.length; j++) {
                hash = ((hash << 5) - hash) + projData.url.charCodeAt(j);
                hash = hash & hash;
            }

            projects.push({
                id: `wkn_${Math.abs(hash)}`,
                platform: 'Workana',
                title: projData.title,
                url: projData.url,
                rewardRaw: projData.rewardRaw,
                rewardMin,
                rewardMax,
                skills: projData.skills,
                full_description,
                status: 'raw',
                scraped_at: new Date().toISOString()
            });

            await detailPage.close();

        } catch (e) {
            console.log(`Erro em ${url}: ${e.message}`);
        }
    }

    if (projects.length > 0) {
        fs.writeFileSync(RAW_FILE, JSON.stringify(projects, null, 2));
        console.log(`\n✅ Sucesso! ${projects.length} Projetos com Deep Fetch salvos em: ${RAW_FILE}`);
    }

    await browser.close();
}

runScraper().catch(console.error);
